File: 2020 年 (25 歳) 設計、デザインパターン/design-patterns-practice/Java 言語で学ぶデザインパターン入門/01 Iterator 一 1 つ 1 つ数え上げる/02 C++/01/BookShelf.cpp
#include "BookShelf.h"

#include <iostream>



BookShelf::BookShelf()
{
	std::cout << __FUNCTION__ << std::endl;
}

BookShelf::~BookShelf()
{
	std::cout << __FUNCTION__ << std::endl;
}

std::unique_ptr<Iterator> BookShelf::iterator() const
{
	return std::make_unique<BookShelfIterator>(*this);
}

File: 2020 年 (25 歳) 設計、デザインパターン/design-patterns-practice/Java 言語で学ぶデザインパターン入門/01 Iterator 一 1 つ 1 つ数え上げる/01 Java/01/BookShelf.java
import java.util.Iterator;



public class BookShelf implements Iterable<String>
{
    private String[] elements_;

    

    public BookShelf(String[] elements_)
    {
        this.elements_ = elements_;
    }

    @Override
    public Iterator<String> iterator()
    {
        return new BookShelfIterator(elements_);
    }
}
